### ISSUE TITLE
Unclear how/where you populate some variables

### ISSUE THREAD

**Issue by OliverBarrett** [amueller/introduction_to_ml_with_python#63]
It's not clear from the snippet notebook approach (I'm using a Python IDE) how you populate variables along the way in your examples.

For instance, with your Ridge Regression example:

from sklearn.linear_model import Ridge

ridge = Ridge().fit(X_train, y_train)
print("Training set score: {:.2f}".format(ridge.score(X_train, y_train)))
print("Test set score: {:.2f}".format(ridge.score(X_test, y_test)))

How did you populate X_train, y_train?

Will this example run as is?

---

My goal, with any book like this is to be able to copy and paste the code and have it run. Now, having said that, I'm now to Python so I may not understand this whole "notebook" thing that seems so prevalent.

I was hoping for individual standalone, working, programs (py files). 

Can you please explain how to get all the examples to work in your notebook where it doesn't "seem" obvious how you are populating the variables?

If I can figure that out, I'll definitely buy your book.

Thanks much.

- O

**Comment by amueller** [comment]
Each chapter is one notebook, which means it has one name-space. That is maybe not the very best of abstractions, but it's also not a bad one, I think.
Why do you not want to load and run the notebooks, instead of using and IDE? that will make it much easier to follow along.

My thought was that having each cell / code block be stand-alone will induce too much redundancy and hinder readability. If I'd do it again, I'd probably split it up into more notebooks, one per subchapter or something like that.

The only way to make sure that variables are defined is to convert each notebook (that is each chapter) to a single ``.py`` file and then run it (you can use nbconvert to convert a notebook to a py file, or open the notebook in jupyter and save it as a py file)

In practice, you shouldn't need to go back more than one or two cells to find the definition of any variable. If you have to go hunting, that's probably a sign that I should have made it more explicit, and I'm happy to take suggestions.

For the example you show, the data is loaded two cells above. In my version you code is ``In[31]``, ``X`` and ``y`` are defined in ``In[29]`` from ``load_extended_boston``.

**Comment by OliverBarrett** [comment]
I am using an IDE, PyCharm. That's kind of where I'm stuck. I would like to see all the relevant code for (I picked one example) Ridge Regression in one place so I don't need to search and figure out how the data get loaded.

I didn't know about converting a notebook to a py file (New to Python). I'll give that a try.

Do you recommend getting your paperback or would the Kindle version possibly be more up to date. I'm guessing the paperback book would be easier to sit and read through before attempting examples.

Thanks for your reply! :)

**Comment by amueller** [comment]
Why are you stuck with an IDE? You can just install jupyter. For developing applications, an IDE is certainly better, but for this, I feel you'd have an easier time with Jupyter.

Unfortunately you have to go an figure out where the data was loaded, though, as I said, that should be really easy. In data analysis and machine learning are an iterative process, and reproducing every previous step for every new step is not really feasible. Though here including the two lines for data loading and splitting would have been an option. In other places in the book, it might be harder, though.

I don't think I have access to the information which print is provided for the kindle. But the ebook is in color (the book is b/w), which might be nice (if you're using a kindle that shows color). The kindle version should also be updated if there's a new print, which the book obviously won't.

**Comment by OliverBarrett** [comment]
Gotcha. Yep, I tried to save the supervised learning notebook as a py file, but it won't run in the IDE without mods.

/Library/Frameworks/Python.framework/Versions/3.6/bin/python3.6 /Users/jim/PycharmProjects/IntroMachineLearning(Mueller)/net/supervisedlearning/one/BookCode.py
Traceback (most recent call last):
  File "/Users/PycharmProjects/IntroMachineLearning/net/supervisedlearning/one/BookCode.py", line 5, in <module>
    get_ipython().magic('matplotlib inline')
AttributeError: 'NoneType' object has no attribute 'magic'

---

Probably just better to read the book and figure things out. If I have a problem I can post here. Getting author support is very helpful and I thank you for your replies.

Being totally new to ML (and Python), I was hoping for a book that would hold my hand a bit as in: type in some code, run some code, (hopefully understand what I just did), etc.

My client is big into SVMs and FDA. I have no idea how to do any of that yet...

Making the transition to Python after 10+ years doing Java. 

Thanks again,

**Comment by amueller** [comment]
If you're new to data science in python, I highly recommend https://github.com/jakevdp/PythonDataScienceHandbook (which is free online) which is a bit more introductory than my book.

**Comment by OliverBarrett** [comment]
Too late. I already ordered your book.:)  However, I do have other reference books too covering NumPy, etc.

Looking through the book, I do understand the examples and after 10+ years doing Java, python, although still a bit strange to me now, will be fine.

I've tried some examples and gotten them working. Time to get the book and read it.

Appreciate your replies. I'll post any problems, but the book is clearly well written at at the right level for me.

-----

BTW, I totally agree with your comment about most ML books. I have stat/math degrees, but although I more or less understand all the notation (it's been a "while" since I took Linear Algebra), those books are worthless to me.

Thanks again for your replies.
